### ISSUE TITLE
next js 13 server error " TypeError: fetch failed "

### ISSUE THREAD

**Issue by AhmedShehata98** [vercel/next.js#54961]
### Link to the code that reproduces this issue or a replay of the bug

https://github.com/AhmedShehata98/shoperz

### To Reproduce

1. Start the application with npm run dev
2. After many re-render or refresh 
3. Facing  500 Internal Server Error page
4. in terminal get this error : 

TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async invokeRequest (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\server-ipc\invoke-request.js:17:12)
    at async invokeRender (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:254:29)
    at async handleRequest (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:447:24)
    at async requestHandler (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:464:13)
    at async Server.<anonymous> (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\start-server.js:117:13) {
  cause: Error: connect ECONNREFUSED ::1:58392
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 58392
  }
}

### Current vs. Expected behavior

i followed the server start instractions by " npm run dev " I expected its works fine without error but got error 
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async invokeRequest (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\server-ipc\invoke-request.js:17:12)
    at async invokeRender (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:254:29)
    at async handleRequest (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:447:24)
    at async requestHandler (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\router-server.js:464:13)
    at async Server.<anonymous> (D:\my-work\Progo-soft\libya-zon\node_modules\next\dist\server\lib\start-server.js:117:13) {
  cause: Error: connect ECONNREFUSED ::1:58392
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 58392
  }
}

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.17.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.19
      eslint-config-next: 13.4.19
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

**Comment by awe79** [comment]
This is not a new issue. This has have been happening since roughly 13.2.x However it seems to have gotten worse since 13.4.19. For me on windows, i'm getting now 1 crash per day on my development environment. In the past the next server would restart which was nice, but now i have to manually restart. Also it seems to be less of a problem on mac. If you search there might me some tips in similar topics. In general development on windows with Next 13 is far from optimal

**Comment by felri** [comment]
I have the same problem, in my case is when I'm sending files from (5 - 20mb) to supabase storage.

```js
const uploadResult = await storageClient
  .from(STORAGE_BUCKET)
  .upload(filePath, file, {
    contentType: `video/${fileExt}`,
    upsert: true
  });
```

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11522:11) {
  cause: Error: write EPIPE
      at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16)
      at WriteWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -32,
    code: 'EPIPE',
    syscall: 'write'
  }
}
```

**Comment by harrisyn** [comment]
Confirming that this has been driving me nuts lately.

really frustrating building in NextJs lately.

**Comment by felri** [comment]
@balazsorban44 I'm using canary btw

`"next": "^13.4.20-canary.15",
`

**Comment by balazsorban44** [comment]
Anyone saying "same issue", or not posting helpful comments to debug, please note that adding a reproduction link helps us verify/investigate more than anything. Stating your Next.js version in itself or an out-of-context code snippet is not very helpful. :pray:  Please comment your reproduction repository instead, we would like to investigate!

I added the label `please verify canary`, because the linked reproduction is on `13.3.1` which closes in on being half a year old version. We've introduced many improvements since then. The reproduction is also pretty complex and is not clear from the reproduction steps how to recreate the issue. @AhmedShehata98 do you have some specific steps you could share?

@cosieLq `undici` is the de facto `fetch` implementation that powers Node.js 18+ versions. In those cases, we don't polyfill at all, just rely on the underlying runtime, so in that case this would indicate an upstream bug. Do you have a codebase we could look at to verify what's the issue that you are seeing "getting worse"?

For anyone, upgrading Node.js might be a good start to verify if it's related or not. :+1:

**Comment by eduardolandaiqmetrix** [comment]
Adding a link to a Git Project which can replicate the same issue, I tested it by using Node  16 and 18, [Git Project](https://github.com/iamshaunjp/nextjs-masterclass/tree/lesson-33), It may require some test supabase credentials to work but does the job to replicate the issue. Error: 

```
"TypeError: fetch failed\n    at Object.fetch (node:internal/deps/undici/undici:11576:11)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
```

And another medium link where a similar issue has been explained and "tried" to fix 
https://medium.com/@kaloyan_17221/fix-vercel-next-js-fetch-failed-from-undici-polyfill-8c66346c9c2f

**Comment by cosieLq** [comment]
@balazsorban44  I don't have a codebase to prove that it is indeed "getting worse". It's just my guess.

I'm working on a project with NextJs v13.2.4. We were trying to upgrade it to v13.4.x. And then in the production environment, the "TypeError: fetch failed" showed up so often (~1 per minute) that we had to roll it back.

Then after a bit searching, I found these issues: #51605 #53353 , thinking that these issues are somehow related to our problem.

In our case, the error codes differ ("ECONNREFUSED" and "ECONNRESET"). They take place after calling on our NextJs API route.

I'm not sure if it's a (upstream) _bug_ or not. It has something to do with load.
If I call API route at high rate (with curl for example), the "TypeError: fetch failed" logs also appear with NextJs v13.2.4. But in production we haven't seen it so far as often as with v13.4.x.

That's why I'm also curious about the performance of NextJs API routes. Are there guidelines hereof?
Because of these errors, we aren't able to upgrade NextJs and leverage the app dir.

**Comment by LufyCZ** [comment]
A reproduction that works consistently for me:

[Repo](https://github.com/sushiswap/sushiswap)
feature/steer-integration

`
pnpm i
export NEXT_PUBLIC_POOLS_API_V0_BASE_URL=https://pools-git-feature-steer-integration.sushi.com
pnpm exec turbo run dev --filter=evm
`

go to [http://localhost:3000/pool/137:0x21988c9cfd08db3b5793c2c6782271dc94749251/smart](http://localhost:3000/pool/137:0x21988c9cfd08db3b5793c2c6782271dc94749251/smart)

open the file
`sushiswap/apps/evm/ui/pool/Steer/SteerLiquidityDistributionWidget/SteerLiquidityInRangeChip.tsx`

change `Range` on line 47 to `Rangea` and back to `Range` a couple times, it's happens faster one time and slower the other

Might have something to do with a client component under a server component? Wild guess.

Noticed this line in the log a couple seconds before it started erroring out:

**Comment by eduardolandaiqmetrix** [comment]
I tried yesterday the last Canary version while using Node 18 and it worked.

**Comment by SpBills** [comment]
For some reason my repository random stalled out and started having this issue on `dev`. Haven't been able to get it to work since.

**Comment by harrisyn** [comment]
@balazsorban44 I wish I could provide more details, I basically found it difficult to work in dev mode.
I had upgraded the next version from 13.4.10 to 13.4.19, and would consistently get the above error, and my dev environment became unbearably slow.

I downgraded to 13.4.13 and it's been stable since.

I might test out the canary version once I roll my current implementation out.

hopefully, someone else may have provided more context before then.

**Comment by gregawoods** [comment]
Hi,

Just wanted to chime in to say we're experiencing the same symptoms in our application. We recently upgraded next.js from 13.4.7 to 13.4.19. Node version 18.7.1.

We see steady increase in memory usage culminating in a `The server is running out of memory, restarting to free up memory.` log message. After that the app starts returning an internal server error and logging out the same `TypeError: fetch failed`  stack trace as above over and over. Requires a restart to get the app running again.

I'm going to try rolling back next.js and let it run for a day or two.

**Comment by meotimdihia** [comment]
This problem happens with `next start` on the production. 
It happens randomly when our server is busy, once it happens next.js server can't continue to work => We have to restart the Next.js server.

**Comment by madfcat** [comment]
Updated from 13.4.10 to 13.4.19 and now get this error

**Comment by karlhorky** [comment]
This issue seems pretty similar:

- https://github.com/vercel/next.js/issues/49896

**Comment by Elvincth** [comment]
This issue continues happened with the latest ^13.4.19 version with my m1 machine in development mode, it happens randomly.

**Comment by karlhorky** [comment]
For all the people who have been reporting that it happens with Next.js v13.4.19, please upgrade to the latest canary on [the Releases page](https://github.com/vercel/next.js/releases) (currently `13.4.20-canary.23`) and try again - this will be helpful to see if the problem still occurs

(I am also experiencing this issue on the latest canary personally, but would be good to get some other data points)

**Comment by Elvincth** [comment]
Hello there! I recently switched to version 13.4.20-canary.23 and encountered an unexpected type error that read "X TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed". I'm unsure of what caused this error, as there is no backend portion to my next.js project. 

<img width="1377" alt="image" src="https://github.com/vercel/next.js/assets/6602723/5f7f0cac-398b-4003-b09b-ca5ac86a5e27">

**Comment by LiorDueto** [comment]
cross-posting in case it helps:
We also encountered similar 'out of memory', 'fetch' errors and eventual crashes (13.4.19).
[Disabling image optimization](https://nextjs.org/docs/app/api-reference/components/image#unoptimized) solved this for us and server memory in production has become more stable.
Just add `images: { unoptimized: true}` to next configuration.

Eventually we do want to benefit from image optimization, but at this point the memory footprint is just too much.

**Comment by naolmelesse** [comment]
> For all the people who have been reporting that it happens with Next.js v13.4.19, please upgrade to the latest canary on [the Releases page](https://github.com/vercel/next.js/releases) (currently `13.4.20-canary.23`) and try again - this will be helpful to see if the problem still occurs
> 
> (I am also experiencing this issue on the latest canary personally, but would be good to get some other data points)

Hello @karlhorky , I have been facing the same issue with Next.js v13.4.19, (Node.js v18.16.0). it is happening frequently. It restarts in less than 5 minutes of starting the server. I am new to Next.js and How do I upgrade to the canary release?

**Comment by Elvincth** [comment]
> > For all the people who have been reporting that it happens with Next.js v13.4.19, please upgrade to the latest canary on [the Releases page](https://github.com/vercel/next.js/releases) (currently `13.4.20-canary.23`) and try again - this will be helpful to see if the problem still occurs
> > (I am also experiencing this issue on the latest canary personally, but would be good to get some other data points)
> 
> Hello @karlhorky , I have been facing the same issue with Next.js v13.4.19, (Node.js v18.16.0). it is happening frequently. It restarts in less than 5 minutes of starting the server. I am new to Next.js and How do I upgrade to the canary release?

See [here](https://nextjs.org/docs/messages/version-staleness) for updating to canary release.

**Comment by eashish93** [comment]
For anyone who is getting this issue, the workaround is to use next@13.4.12 . Another workaround and why this is happening is because fetch url protocol converted to `https` on production with localhost. You can reproduce this in dev env using `cloudflared`: 
- Run `pnpm dev` one tab
- Run `cloudflared tunnel --url "<localhost-url>"
- Add log for `req.nextUrl.origin` somewhere in `middleware.ts` file. 
- Browse cloudflared tunnel url and check the log, you'll see `https://localhost:...` which is the cause of this error if you're using `req.nextUrl` anywhere in server fetch.
So to solve this either next@13.4.12 or use `http:${req.nextUrl.host}` as temporary workaround.

**Comment by luistak** [comment]
> I recently switched to version 13.4.20-canary.23 and encountered an unexpected type error that read "X TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed". I'm unsure of what caused this error, as there is no backend portion to my next.js project.
> 
> <img alt="image" width="1377" src="https://user-images.githubusercontent.com/6602723/267101947-5f7f0cac-398b-4003-b09b-ca5ac86a5e27.png">

Same here, I'm having an issue with high memory consumption in a K8s environment using the app router where regardless of the amount of memory I set on a pod it always hits 95 ~ 97% memory usage, I tried disabling image optimization, removing sharp, upgrading node [as recommended](https://github.com/vercel/next.js/issues/49929#issuecomment-1691269445) but only after upgrading next to `13.4.20-canary.23` the memory usage got better to 80 ~ 85% usage although it crashed several `fetch` requests and unexpected errors like this one so I rollbacked to `13.4.19` no more errors appeared

**Comment by vikyw89** [comment]
> "next": "13.4.20-canary.24"

updated to this version, haven't experienced another crash yet and dev reload is slightly faster.

here's my deps
  "dependencies": {
    "@akebifiky/remark-simple-plantuml": "^1.0.2",
    "@auth/firebase-adapter": "^1.0.0",
    "@auth/prisma-adapter": "^1.0.2",
    "@reduxjs/toolkit": "^1.9.5",
    "@types/node": "20.5.7",
    "@types/react": "18.2.21",
    "@types/react-dom": "18.2.7",
    "autoprefixer": "10.4.15",
    "eslint": "8.48.0",
    "eslint-config-next": "13.4.19",
    "firebase": "^10.3.1",
    "firebase-admin": "^11.10.1",
    "mermaid": "^10.4.0",
    "mermaid.cli": "^0.3.6",
    "next": "13.4.20-canary.24",
    "next-auth": "^4.23.1",
    "postcss": "8.4.29",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-icons": "^4.10.1",
    "react-markdown": "^8.0.7",
    "react-redux": "^8.1.2",
    "rehype-raw": "^7.0.0",
    "remark-gfm": "^3.0.1",
    "remark-mermaid": "^0.2.0",
    "remark-mermaid-plugin": "^1.0.2",
    "zod": "^3.22.2"
  },
  "devDependencies": {
    "@tailwindcss/typography": "^0.5.10",
    "@typescript-eslint/parser": "^6.5.0",
    "daisyui": "^3.6.4",
    "eslint-config-prettier": "^9.0.0",
    "prettier": "3.0.3",
    "prettier-plugin-tailwindcss": "^0.5.4",
    "tailwind-scrollbar": "^3.0.5",
    "tailwindcss": "^3.3.3",
    "typescript": "^5.2.2"
  },

**Comment by antmonnier** [comment]
Seems identical to https://github.com/vercel/next.js/issues/49578. @b1rdex provided a reproduction [here](https://github.com/b1rdex/next.js-fetch-error).

**Comment by Elvincth** [comment]
Another new memory error while using with next.js version `13.4.20-canary.26`

```

<--- Last few GCs --->

[2248:0x138040000]  9308182 ms: Scavenge 3768.4 (3885.6) -> 3767.5 (3885.6) MB, 4.5 / 0.0 ms  (average mu = 0.315, current mu = 0.313) allocation failure; 
[2248:0x138040000]  9308191 ms: Scavenge 3768.7 (3885.6) -> 3768.0 (3885.6) MB, 5.6 / 0.0 ms  (average mu = 0.315, current mu = 0.313) allocation failure; 
[2248:0x138040000]  9309636 ms: Mark-sweep 3768.2 (3885.6) -> 3764.7 (3889.0) MB, 1444.8 / 0.0 ms  (average mu = 0.191, current mu = 0.042) allocation failure; GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1023a8a88 node::Abort() [/Users//.nvm/versions/node/v18.16.1/bin/node]
 2: 0x1023a8c78 node::ModifyCodeGenerationFromStrings(v8::Local<v8::Context>, v8::Local<v8::Value>, bool) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 3: 0x1024fe548 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 4: 0x1026a8e00 v8::internal::EmbedderStackStateScope::EmbedderStackStateScope(v8::internal::Heap*, v8::internal::EmbedderStackStateScope::Origin, cppgc::EmbedderStackState) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 5: 0x1026ac9ec v8::internal::Heap::CollectSharedGarbage(v8::internal::GarbageCollectionReason) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 6: 0x1026a9a00 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*, v8::GCCallbackFlags) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 7: 0x1026a6d00 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 8: 0x10269b83c v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users//.nvm/versions/node/v18.16.1/bin/node]
 9: 0x10268134c v8::internal::Factory::CodeBuilder::AllocateCode(bool) [/Users//.nvm/versions/node/v18.16.1/bin/node]
10: 0x102680c6c v8::internal::Factory::CodeBuilder::BuildInternal(bool) [/Users//.nvm/versions/node/v18.16.1/bin/node]
11: 0x10338455c v8::internal::compiler::CodeGenerator::FinalizeCode() [/Users//.nvm/versions/node/v18.16.1/bin/node]
12: 0x10356bc20 void v8::internal::compiler::PipelineImpl::Run<v8::internal::compiler::FinalizeCodePhase>() [/Users//.nvm/versions/node/v18.16.1/bin/node]
13: 0x1035629b0 v8::internal::compiler::PipelineImpl::FinalizeCode(bool) [/Users//.nvm/versions/node/v18.16.1/bin/node]
14: 0x103562820 v8::internal::compiler::PipelineCompilationJob::FinalizeJobImpl(v8::internal::Isolate*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
15: 0x1025b3f5c v8::internal::Compiler::FinalizeTurbofanCompilationJob(v8::internal::TurbofanCompilationJob*, v8::internal::Isolate*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
16: 0x1025d5524 v8::internal::OptimizingCompileDispatcher::InstallOptimizedFunctions() [/Users//.nvm/versions/node/v18.16.1/bin/node]
17: 0x102653910 v8::internal::StackGuard::HandleInterrupts() [/Users//.nvm/versions/node/v18.16.1/bin/node]
18: 0x102a21db4 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
19: 0x102d7104c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users//.nvm/versions/node/v18.16.1/bin/node]
20: 0x102d481c4 Builtins_MapPrototypeSet [/Users//.nvm/versions/node/v18.16.1/bin/node]
21: 0x10f11f360 
22: 0x10f11b620 
23: 0x108852684 
24: 0x10f11f48c 
25: 0x10f21efe4 
26: 0x10f11f48c 
27: 0x10f11b620 
28: 0x109770370 
29: 0x10f11f48c 
30: 0x10f21e318 
31: 0x10977acfc 
32: 0x10976d054 
33: 0x107db0ebc 
34: 0x10f5968b8 
35: 0x10f59690c 
36: 0x1095540b4 
37: 0x10976d054 
38: 0x107db0ebc 
39: 0x1097779b4 
40: 0x102d2def4 Builtins_AsyncFunctionAwaitResolveClosure [/Users//.nvm/versions/node/v18.16.1/bin/node]
41: 0x102dbc738 Builtins_PromiseFulfillReactionJob [/Users//.nvm/versions/node/v18.16.1/bin/node]
42: 0x102d1fc4c Builtins_RunMicrotasks [/Users//.nvm/versions/node/v18.16.1/bin/node]
43: 0x102cfa3a4 Builtins_JSRunMicrotasksEntry [/Users//.nvm/versions/node/v18.16.1/bin/node]
44: 0x10262ac4c v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users//.nvm/versions/node/v18.16.1/bin/node]
45: 0x10262b13c v8::internal::(anonymous namespace)::InvokeWithTryCatch(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users//.nvm/versions/node/v18.16.1/bin/node]
46: 0x10262b318 v8::internal::Execution::TryRunMicrotasks(v8::internal::Isolate*, v8::internal::MicrotaskQueue*, v8::internal::MaybeHandle<v8::internal::Object>*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
47: 0x102651a80 v8::internal::MicrotaskQueue::RunMicrotasks(v8::internal::Isolate*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
48: 0x10265221c v8::internal::MicrotaskQueue::PerformCheckpoint(v8::Isolate*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
49: 0x1022f0c4c node::InternalCallbackScope::Close() [/Users//.nvm/versions/node/v18.16.1/bin/node]
50: 0x1022f0fd0 node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/Users//.nvm/versions/node/v18.16.1/bin/node]
51: 0x102305c58 node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
52: 0x1023ad8c8 node::fs::FSReqCallback::Reject(v8::Local<v8::Value>) [/Users//.nvm/versions/node/v18.16.1/bin/node]
53: 0x1023ae154 node::fs::FSReqAfterScope::Reject(uv_fs_s*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
54: 0x1023ae388 node::fs::AfterNoArgs(uv_fs_s*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
55: 0x1023a49f8 node::MakeLibuvRequestCallback<uv_fs_s, void (*)(uv_fs_s*)>::Wrapper(uv_fs_s*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
56: 0x102cd72a0 uv__work_done [/Users//.nvm/versions/node/v18.16.1/bin/node]
57: 0x102cdaa5c uv__async_io [/Users//.nvm/versions/node/v18.16.1/bin/node]
58: 0x102ced010 uv__io_poll [/Users//.nvm/versions/node/v18.16.1/bin/node]
59: 0x102cdaf2c uv_run [/Users//.nvm/versions/node/v18.16.1/bin/node]
60: 0x1022f16e0 node::SpinEventLoop(node::Environment*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
61: 0x1023e5dd4 node::NodeMainInstance::Run() [/Users//.nvm/versions/node/v18.16.1/bin/node]
62: 0x102375ab0 node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult const*) [/Users//.nvm/versions/node/v18.16.1/bin/node]
63: 0x102375d68 node::Start(int, char**) [/Users//.nvm/versions/node/v18.16.1/bin/node]
64: 0x192413f28 start [/usr/lib/dyld]
```

**Comment by karlhorky** [comment]
@Elvincth that seems unrelated to this `fetch` error, but in your stack trace it's showing Node.js v18.16.1 - there's a memory leak before Node.js v18.17.1, so you should upgrade :)

**Comment by k2xl** [comment]
Note we saw this on 13.4.19 (when we upgraded from 13.3.4). Downgrading to 13.4.13 has the same issue for us actually. Currently trying 13.4.10.

Seems like it takes a few hours before it begins to spam that fetch error in the logs and return internal service...

**Comment by A7med3bdulBaset** [comment]
I have a similar issue. I'm building a front-end project in Next.js and the backend is in Django. When I send a request from a client component I get `fetch failed` and When I move the fetch to server action or route handler it works fine.

**Comment by owenallenaz** [comment]
@k2xl That is the similar experience for us. The thing that makes it extremely difficult to diagnose is that I can spam requests on my local run instance (running prod settings), and it never leaks an ounce of memory, but I also don't see these `fetch failed`, they only appear after the server has been running in production while under significant load. They appear to be coming from some sort of internal subsystem but the stack trace is utterly useless. I've analyzed my application thoroughly and that `fetch` which is erroring is simply not from a fetch in our code, it's coming from within Next. If we knew where, then it would be easier to try and track down the memory leak.

**Comment by VCasecnikovs** [comment]
I have solved this problem for myself:

The problem happens, when I send File to Supabase Storage.

To solve the problem, I send Array Buffer:
```
let array_buffer = await image.arrayBuffer();
```

**Comment by meotimdihia** [comment]
> This problem happens with `next start` on the production. Ubuntu 20.20 Node.js v18.17.1 It happens randomly when our server is busy, once it happens next.js server can't continue to work => We have to restart the Next.js server.
> 
> Fastify servers work fine, but it just happens with Next.js servers.
> 
> P/S Downgrade to `"next": "13.4.12"` works fine for me.

I tried to update to  `"next": "13.4.20-canary.33"`, and it looks like the problem won't happen again.

**Comment by PeterJSung** [comment]
@meotimdihia This issue is really weird. We do not have this problem after downgrading to Next 13.4.12 version. Some people recommend downgrading to 13.4.7.

The cause of the issue seems to be the undici issue in the link below. 
https://github.com/nodejs/undici/issues/1602
It doesn't seem to have been resolved yet, but I'm glad to hear that the problem didn't recur in Next 13.4.20.

**Comment by karlhorky** [comment]
> I updated next.js to `13.4.20-canary.33`, and it looks like the problem was fixed.

I upgraded to [`next@13.4.20-canary.35`](https://github.com/vercel/next.js/releases/tag/v13.4.20-canary.35) and we have also not experienced the `fetch` failures since then... 👀 🤔 

I will keep an eye on this, but maybe this has been fixed

**Comment by diegohaz** [comment]
Still seeing some random `TypeError: fetch failed` in `v13.5.1`.

**Comment by amesas** [comment]
Same prob with v13.5.1, going back to 13.4.12 :(

**Comment by harrisyn** [comment]
updated to v13.5.1 a couple of hours ago, and npm run build fails consistently with issues.

Rolling back to 13.4.12

Error: NextRouter was not mounted. https://nextjs.org/docs/messages/next-router-not-mounted
    at p (/app/.next/server/chunks/4914.js:1:14391)
    at f (/app/.next/server/chunks/7648.js:1:487)
    at renderWithHooks (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5658:16)
    at renderIndeterminateComponent (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5731:15)
    at renderElement (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5946:7)
    at renderNodeDestructiveImpl (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:6104:11)
    at renderNodeDestructive (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:6076:14)
    at renderIndeterminateComponent (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5785:7)
    at renderElement (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5946:7)
    at renderNodeDestructiveImpl (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:6104:11)

Error occurred prerendering page "/superadmin". Read more: https://nextjs.org/docs/messages/prerender-error
Error: NextRouter was not mounted. https://nextjs.org/docs/messages/next-router-not-mounted
    at p (/app/.next/server/chunks/4914.js:1:14391)
    at f (/app/.next/server/chunks/7648.js:1:487)
    at renderWithHooks (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5658:16)
    at renderIndeterminateComponent (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5731:15)
    at renderElement (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5946:7)
    at renderNodeDestructiveImpl (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:6104:11)
    at renderNodeDestructive (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:6076:14)
    at renderIndeterminateComponent (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5785:7)
    at renderElement (/app/node_modules/react-dom/cjs/react-dom-server.browser.development.js:5946:7)
 ✓ Generating static pages (51/51)

**Comment by meotimdihia** [comment]
@harrisyn I think your problem is not related to this thread.

**Comment by harrisyn** [comment]
> @harrisyn I think your problem is not related to this thread.

maybe not directly, however I only updated to v13.5.1 instead of canary to get rid of this fetch failed issue.

**Comment by karlhorky** [comment]
@harrisyn better to either [create a new issue](https://github.com/vercel/next.js/issues/new?assignees=&labels=template%3A+bug&projects=&template=1.bug_report.yml) or check for existing issues, the one below sounds similar to yours:

- https://github.com/vercel/next.js/issues/53045

**Comment by pealmeid** [comment]
I'm getting this same error after upgrading to `13.5.3` from `13.5.2`. I'm getting the following error:

`TypeError: fetch failed`
`    at Object.fetch (node:internal/deps/undici/undici:11576:11)`
`    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {`
`  cause: Error: self-signed certificate`
`      at TLSSocket.onConnectSecure (node:_tls_wrap:1600:34)`
`      at TLSSocket.emit (node:events:517:28)`
`      at TLSSocket._finishInit (node:_tls_wrap:1017:8)`
`      at ssl.onhandshakedone (node:_tls_wrap:803:12)`
`      at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17) {`
`    code: 'DEPTH_ZERO_SELF_SIGNED_CERT'`
`  }`
`}`

Our solution proxies calls to an external https API server (that's run locally in the dev environment).

We were getting a similar error in the past and our investigation pointed to extending Node's local certificates by using the `NODE_EXTRA_CA_CERTS` environment variable with the path to a self-signed certificate. After setting this variable the error stopped happening but now it has returned.

Note: if we downgrade to Next `13.5.2` we don't get any errors. The only difference in `package.json` between the working and the failing solution is Next's version.

Thanks in advance for any guidance on this.

**Comment by GoudekettingRM** [comment]
Also getting this. First on `13.4.19`, upgraded to `13.5.3`, but still getting it. 

The following two errors alternate in my console:
```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async imageOptimizer (/Users/.../node_modules/next/dist/server/image-optimizer.js:521:29)
    at async cacheEntry.imageResponseCache.get.incrementalCache (/Users/.../node_modules/next/dist/server/next-server.js:519:61)
    at async /Users/.../node_modules/next/dist/server/response-cache/index.js:102:36 {
  cause: Error: connect ENETUNREACH 64:ff9b::34da:19d3:443
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -51,
    code: 'ENETUNREACH',
    syscall: 'connect',
    address: '64:ff9b::34da:19d3',
    port: 443
  }
}
```
```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async imageOptimizer (/Users/.../node_modules/next/dist/server/image-optimizer.js:521:29)
    at async cacheEntry.imageResponseCache.get.incrementalCache (/Users/.../node_modules/next/dist/server/next-server.js:519:61)
    at async /Users/.../node_modules/next/dist/server/response-cache/index.js:102:36 {
  cause: ConnectTimeoutError: Connect Timeout Error
      at onConnectTimeout (/Users/.../node_modules/next/dist/compiled/undici/index.js:1:92227)
      at /Users/.../node_modules/next/dist/compiled/undici/index.js:1:91719
      at Immediate._onImmediate (/Users/.../node_modules/next/dist/compiled/undici/index.js:1:92109)
      at process.processImmediate (node:internal/timers:476:21)
      at process.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'UND_ERR_CONNECT_TIMEOUT'
  }
}
```

Only on `next/image`, no errors when just using `img` tags (obviously ... as those don't use `imageOptimizer`).


I temporarily resolved it by just skipping the default image loader, but that is not the long term solution. I want the default loader for it's optimisations haha

```js
// next.config.js
...
images: {
    loader: 'custom',
    loaderFile: './bin/imageLoader.js',
},
...

// imageLoader.js
'use client';

export default function imageLoader({ src, width, quality }) {
  if (src.includes('?')) {
    return `${src}&w=${width}&q=${quality || 75}`;
  }

  return `${src}?w=${width}&q=${quality || 75}`;
}
```

**Comment by meotimdihia** [comment]
@GoudekettingRM Downgrade to 13.5.2 => It should be fine.
13.5.3 introduced a new bug with fetch - I think.

**Comment by GoudekettingRM** [comment]
@meotimdihia 

Thanks for the quick reply,  same issues on 13.5.2 unfortunately...

**Comment by thany** [comment]
Still the same on 13.5.3.

Revert to 13.4.x works around it, but it had a [severe bug pertaining to locales](https://github.com/vercel/next.js/issues/54217) Until only a few days ago, iirc.

So, it's a matter of choose the lesser evil, at least for the moment 😟

**Comment by dsacramone** [comment]
uggg.. seeing it too: 13.5.2

 ⨯ TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at getStaticProps (webpack-internal:///./src/pages/us/[...location].tsx:119:59) {
  cause: Error: connect ECONNREFUSED ::1:80
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
    errno: -61,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 80
  },

**Comment by joggienl** [comment]
I noticed the same issue as @pealmeid , I have something in place on my system that sets `NODE_EXTRA_CA_CERTS`. I need it to be able to use some local https service.

The cause of this issue is this PR: https://github.com/vercel/next.js/pull/55775. There you can see that `process.env.NODE_EXTRA_CA_CERTS` gets set. I think that should not be done.

A simple reproduction is to put an export in your login shell that sets `NODE_EXTRA_CA_CERTS`. If you use zsh you could add to your ~/.zhsrc a line like:

```
export NODE_EXTRA_CA_CERTS=/path/to/a/ca/you/want/to/load
```

If you put a console.log in next.config.js you can see `process.env.NODE_EXTRA_CA_CERTS` is empty. If you hack into your node_modules and disable the line that sets `process.env.NODE_EXTRA_CA_CERTS` it works again as expected.

I don't think it helps or solves the original problem of this issue, perhaps this should be a separate issue.

**Comment by rcssdy** [comment]
Also following here.. having the same issue as of today, locally. Haven't tried to deploy my project, but have tried all of the workarounds listed here, and still no luck.

**Comment by timneutkens** [comment]
Just read through every post. They're all about different issues and most are not caused by Next.js but by application code.

As a first step upgrade to the latest stable version of Next.js, for example what https://github.com/vercel/next.js/issues/54961#issuecomment-1709991281 and other shared has already been fixed. Though unable to reproduce we were able to just remove that code path altogether so that specifically can't happen anymore.

TLDR: Next.js no longer uses fetch in the hot path while rendering, whereas in 13.4 it did, so this error can't happen in the normal rendering path. 

For others I'm seeing a bunch of different issues being confused:

- `next build` prerender errors out with fetch failed, usually this means:
	- You're trying to fetch an endpoint that doesn't exist at build time (e.g. pages/api or route handlers), for this case call the code directly instead, it's all in the same Next.js application.
	- Your build system doesn't have access to the API (i.e. there's an IP block or something similar)
	- Your build system doesn't have the right API Keys set up (i.e. you forgot to add an environment variable)
- Fetch failed at runtime during rendering
	- I.e. when you `fetch()` in a server component you need to ensure that the url you're fetching is available
	- Your runtime system doesn't have access to the API (i.e. there's an IP block or something similar)
	- Your runtime system doesn't have the right API Keys set up (i.e. you forgot to add an environment variable)
- Fetch failed in Next.js
	- There is a case where Next.js internals `fetch()`, in particular what was mentioned above wrt the image optimizer, however the error shown above is caused by an image url  being provided that fails to fetch, e.g. because you're offline, it's unreachable and such.

Hope that provides enough context!

**Comment by joggienl** [comment]
Thanks for this explanation @timneutkens.

**Comment by thany** [comment]
@timneutkens In the end, if all it takes is to downgrade Next.js to 13.4 (down from 13.5), you can't reasonably assume the problem is in application code. In such case, it doesn't matter if the url `fetch` is trying to access, is blocked or otherwise unavailable. It's demonstrably Next.js just spewing out an error that it didn't before, and is not useful.

Isn't not application code or something getting blocked or whatever else, it is purely Next.js that has put extra unneccesary checks in place for the certificate. And frankly, on a *development* environment, I don't give 2 tosses about the correctness of certificates. If I could work without https, I would, but when I can't, I just don't care about certificates. I'm on a development machine, and I just need to my job instead of messing around with certificates that have always worked perfectly fine. I don't think that's all that unreasonable, that a developer just wants to get his/her work done, and not have to deal with certificate bollocks. Therefor is super unproductive to restrict certificate requirements further than they already were.

**Comment by GoudekettingRM** [comment]
Thanks for the expansive answer @timneutkens.

Not quite sure about the wrt optimizer though. If I skip it, the images load without error - i.e. the images exist and are reachable. As soon as I use the default optimizer it errors with `ENETUNREACH`/`UND_ERR_CONNECT_TIMEOUT`. I understand that that is a message that indicates the optimizer cannot get the image, but seeing that the image is reachable and present, I'm confused as to why not and I'm not convinced it's purely application code error. I've tried a few times and I see that sometimes - after like 5 refreshes - it does load, but not always. If I have some time later today I'll see if I can replicate it in a new project.

**Comment by amesas** [comment]
> Thanks for the expansive answer @timneutkens.
> 
> Not quite sure about the wrt optimizer though. If I skip it, the images load without error - i.e. the images exist and are reachable. As soon as I use the default optimizer it errors with `ENETUNREACH`/`UND_ERR_CONNECT_TIMEOUT`. I understand that that is a message that indicates the optimizer cannot get the image, but seeing that the image is reachable and present, I'm confused as to why not and I'm not convinced it's purely application code error. I've tried a few times and I see that sometimes - after like 5 refreshes - it does load, but not always. If I have some time later today I'll see if I can replicate it in a new project.

Same problem here.

It's a production build, running on windows using a corporate proxy.

To setup the proxy we were setting up HTTPS_PROXY env var 

As it was not enough, ignore in some internal requests, we use node --require to preload some code.

In this case globalAgent.

It worked ok, until undici got into the scene and needed to use the same preloaded code to setup the proxyagent on the global undici object.

And now, since 13.4.13, we get the ECONN_REFUSED on the image optimizer requests. Just downgrading  after 13.4.12, everything works ok.

So it looks like NextJs, has problems with corporate proxies as we always have to look for workarounds

PS: all custom code requests works, just internal requests done by NextJS, fail. So probably our scenario is not on the list by @timneutkens

**Comment by owenallenaz** [comment]
@timneutkens You're evaluation on bullet point 3 I think is just wrong. There are numerous comments here and in the previous thread describing this issue, and it **isn't coming from our code**. I repeat. We never see these fetch failed while we hit our own pages... they appear in the logs later, often after the process has been up for hours. In our specific example I have tried load testing the application for an hour handling hundreds of thousands of requests and there are 0 fetch failed. Yet, our production logs are absolutely full of them, when I'm hitting the same exact URL.

There is some sort of internal mechanic going on which is triggering the fetch failed. It aligns with what many of the commentators here are bringing up, and unfortunately I have zero confidence the Next team even thinks this is an issue.

**Comment by timneutkens** [comment]
Like I said **most** are caused by application code, not **all**. If you're still running into this after upgrading please provide a full reproduction so that we can investigate it further, unfortunately there is no runnable reproduction provided in this issue so far besides the ones mentioning the stacktrace with code that no longer exists.

As said the majority of comments posted on this issue cannot happen anymore as we got rid of the code path being referenced, it no longer exists in 13.5.

Looking forward to looking at a reproduction.

**Comment by timneutkens** [comment]
@harrisyn others already replied to your case , which is to open a new issue with a reproduction: https://github.com/vercel/next.js/issues/54961#issuecomment-1727892539

We're unable to help in the majority of cases where the only information shared is a stack trace, please provide a full repository that we can run to investigate.

**Comment by PeterJSung** [comment]
@timneutkens 
hello. Thank you for your reply. I don't know what happens to other servers, but the above error occurs irregularly for us. Even pages that were previously successful suddenly show errors.

Also, this error occurs starting from a certain version.
Some people say it does not occur starting with v13.4.9, but most people say it does not occur starting with v13.4.12 (the error occurs starting with v13.4.13).

If it occurred starting from a specific version of the same code, it would be difficult to say for sure that it was a problem with the users' application code.

When looking at the differences between versions as shown in the link below, it appears that undici has been applied starting from v13.4.13.
(Issue link https://github.com/vercel/next.js/compare/v13.4.12...v13.4.13#diff-abd1f3c858290361038c2f039d2ebdd031900f46daec6c4a613450dd8fa21dc3)

Also, it appears that a similar issue has been open in undici for quite some time. (issue link https://github.com/nodejs/undici/issues/1602)

From the user's perspective, there is no way to prepare for the sudden occurrence of an error when the patch version is updated from the semantic version.
If backward compatibility is not properly supported, it appears that at least a change is needed starting from the minor version.

We are sorry that we have difficulty providing you with our product code. Probably, it would be difficult for most companies to provide service codes.

**Comment by karlhorky** [comment]
For our apps [the `fetch` error went away with `next@13.4.20-canary.35`](https://github.com/vercel/next.js/issues/54961#issuecomment-1725629311).

But I realize that we have a non-standard configuration for [the Next.js custom server `hostname` option](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server#:~:text=The%20hostname%20the%20server%20is%20running%20behind) (as shown below, we're using `127.0.0.1` instead of `localhost`, which can be found in various issues to be causing connection problems).

Can't promise for sure that it will help, but maybe worth a shot to try to configure the hostname.

```diff
const app = next({
-  hostname: 'localhost',
+  // Use '127.0.0.1' instead of 'localhost' to temporarily work around issues
+  // with ECONNREFUSED (connection refused) errors with high port numbers
+  // https://github.com/vercel/next.js/issues/51684#issuecomment-1609345705
+  //
+  // TODO: Revert to 'localhost' when the issue is resolved https://github.com/vercel/next.js/issues/51684
+  hostname: '127.0.0.1',
```

- Original comment: https://github.com/vercel/next.js/issues/51684#issuecomment-1609668578

**Comment by DuCanhGH** [comment]
@karlhorky I can't reproduce this issue with the latest Next.js canary.
![image](https://github.com/vercel/next.js/assets/75556609/b9965b7e-d70b-495a-9d37-5a5d7acda256)
![image](https://github.com/vercel/next.js/assets/75556609/e32a7a9b-2a70-4ecb-bf06-8d8858c9ef63)

Perhaps updating would solve the problem? Seems that this issue was happening because Next.js' IPv6 handling was... weird back then.

Also, just saying, but I don't see how `127.0.0.1` is a non-standard thing :)

**Comment by karlhorky** [comment]
> I can't reproduce this issue with the latest Next.js canary

Nice, removing the `hostname` configuration wasn't working back when https://github.com/vercel/next.js/issues/49677 was resolved, but I can confirm that removing `hostname` or configuring it with `hostname: 'localhost'` is now working with our apps (on `next@13.5.4`) 🙌 

I'll keep an eye out for any intermittent `fetch` crashes in the next days too, just in case.

**Comment by DuCanhGH** [comment]
@karlhorky I'd say that the fix for that issue and some PRs after that screwed up Next.js' IPv6 handling even further, so yeah. Got fixed real good after that though, should work properly starting from 13.5 :) I fixed that IPv6 handling, but there were some issues after I got that in (13.4.16). They got rid of the render workers after that as well (13.4.20-canary.24).

**Comment by balazsorban44** [comment]
Tim has given a thorough explanation and what to provide/check in https://github.com/vercel/next.js/issues/54961#issuecomment-1743687578 

After that comment, all the following comments were talking about 13.4, while it was pointed out to check with 13.5

The OP's issue has been fixed, as the whole code path of that source of issue has been removed altogether.

There might be mixed problems in this issue (some _might_ be bugs), but those should have a dedicated bug report, with a minimal reproduction that we can investigate! (Keep in mind that `TypeError: fetch failed` is too generic to call it the "same issue". That's why we are asking for reproductions for those as well.)

We understand if you cannot provide your company's source code, but that's not what we were asking for either, as it might not be the best reproduction anyway. Please read: https://github.com/vercel/next.js/issues/54961#issuecomment-1745781190

I will close this issue now, but feel free to open a new one - after testing with the latest version of Next.js - with something that we can investigate, we would like to work on all verified Next.js bugs! :pray:
